### ISSUE TITLE
Support dataclasses in _extract_batch_size

### ISSUE THREAD

**Issue by twsl** [PyTorchLightning/pytorch-lightning#12572]
## 🐛 Bug

As raised on Slack and noticed by me as well, 
https://github.com/PyTorchLightning/pytorch-lightning/blob/44e3edbf7a3fffd3069e2bc24c6ee143de79944b/pytorch_lightning/utilities/data.py#L40-L53

doesnt support dataclasses, which are often used for type checking.

### To Reproduce

Wrap tensors in a dataclass.

### Expected behavior

Extract batchsize correctly.
During logging, the batch_size can be manually specified as a temporary wrokaround.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.6.0
- PyTorch Version (e.g., 1.10): 1.11
- Python version (e.g., 3.9): 3.9
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source): pip
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context
cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7

**Comment by mayeroa** [comment]
One other use case: if the batch is a dataclass containing a list of dataclasses like so:

```python
from dataclasses import dataclass, field
from typing import List

import torch
from pytorch_lightning.utilities.data import extract_batch_size
from torch import Tensor


@dataclass
class TensorDatapoint:
    image: Tensor = torch.randint(0, 255, size=(3, 32, 32))


@dataclass
class Batch:
    items: List[TensorDatapoint] = field(default_factory=lambda: [])


if __name__ == '__main__':
    datapoint = TensorDatapoint()
    batch = Batch(items=[datapoint, datapoint, datapoint, datapoint])
    print(extract_batch_size(batch))
```

I do not know if the proposed solution handles this case but it would be interesting to support that scenario as well.

Currently, it outputs the following error:

```console
pytorch_lightning.utilities.exceptions.MisconfigurationException: We could not infer the batch_size from the batch. Either simplify its structure or provide the batch_size as `self.log(..., batch_size=batch_size)`.
```

Would expect: 

```python
>>> 4
```

**Comment by mayeroa** [comment]
One other use case: if the batch is a dataclass containing a list of dataclasses like so:

```python
from dataclasses import dataclass, field
from typing import List

import torch
from pytorch_lightning.utilities.data import extract_batch_size
from torch import Tensor


@dataclass
class TensorDatapoint:
    image: Tensor = torch.randint(0, 255, size=(3, 32, 32))


@dataclass
class Batch:
    items: List[TensorDatapoint] = field(default_factory=lambda: [])


if __name__ == '__main__':
    datapoint = TensorDatapoint()
    batch = Batch(items=[datapoint, datapoint, datapoint, datapoint])
    print(extract_batch_size(batch))
```

I do not know if the proposed solution handles this case but it would be interesting to support that scenario as well.

Currently, it outputs the following error:

```console
pytorch_lightning.utilities.exceptions.MisconfigurationException: We could not infer the batch_size from the batch. Either simplify its structure or provide the batch_size as `self.log(..., batch_size=batch_size)`.
```

Would expect a batch size of 4 (since four datapoints into the list): 

```python
>>> 4
```

If I understand correctly, the proposed solution would output 3.

**Comment by twsl** [comment]
The current implementation resolves iteratively untill a tensor is resolved and the first dimension of the tensor is used as batch dim.
I think, this point you raised should be handled in a different issue. I would personally replace the implementation in the `Batch` items with a the `torch.stack` of `TensorDatapoint` image
